<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction_to_Linux - Scripts</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Introduction_to_Linux</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./00-intro.html">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./01-start.html">
 <span class="menu-text">Getting Started</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./02-filesystem.html">
 <span class="menu-text">Filesystem</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./03-files.html">
 <span class="menu-text">Files</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./04-processes.html">
 <span class="menu-text">Processes</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./05-analysis.html">
 <span class="menu-text">Analysis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./06-scripts.html" aria-current="page">
 <span class="menu-text">Scripts</span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#shell-scripts-in-bash" id="toc-shell-scripts-in-bash" class="nav-link active" data-scroll-target="#shell-scripts-in-bash">Shell scripts in bash</a>
  <ul class="collapse">
  <li><a href="#writing-and-running-your-own-scripts" id="toc-writing-and-running-your-own-scripts" class="nav-link" data-scroll-target="#writing-and-running-your-own-scripts">Writing and running your own scripts</a></li>
  <li><a href="#editing-scripts" id="toc-editing-scripts" class="nav-link" data-scroll-target="#editing-scripts">Editing scripts</a></li>
  </ul></li>
  <li><a href="#reproducible-data-analysis-with-scripts" id="toc-reproducible-data-analysis-with-scripts" class="nav-link" data-scroll-target="#reproducible-data-analysis-with-scripts">Reproducible data analysis with scripts</a>
  <ul class="collapse">
  <li><a href="#creating-a-simple-data-analysis-script" id="toc-creating-a-simple-data-analysis-script" class="nav-link" data-scroll-target="#creating-a-simple-data-analysis-script">Creating a simple data analysis script</a></li>
  <li><a href="#creating-an-improved-data-analysis-script" id="toc-creating-an-improved-data-analysis-script" class="nav-link" data-scroll-target="#creating-an-improved-data-analysis-script">Creating an improved data analysis script</a></li>
  <li><a href="#managing-your-scripts" id="toc-managing-your-scripts" class="nav-link" data-scroll-target="#managing-your-scripts">Managing your scripts</a></li>
  </ul></li>
  <li><a href="#closing-thoughts-practical-workflows-on-the-command-line" id="toc-closing-thoughts-practical-workflows-on-the-command-line" class="nav-link" data-scroll-target="#closing-thoughts-practical-workflows-on-the-command-line">Closing thoughts: practical workflows on the command line</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Scripts</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>


<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>

<hr>
<p>So far in this course we have learned how to work interactively on the command line to analyse data. This section will introduce the idea of <em>bash shell scripting</em>. We will also introduce the concept of reproducible data analysis, and will show how using scripts to analyse data can facilitate this.</p>
<section id="shell-scripts-in-bash" class="level2">
<h2 class="anchored" data-anchor-id="shell-scripts-in-bash">Shell scripts in bash</h2>
<div class="key-points">
<h2 class="anchored">
<i class="fas fa-thumbtack"></i> Key points
</h2>
<ul>
<li>A <strong>shell script</strong> is a text file containing multiple commands, which can then be run from the command line as a single command</li>
</ul>
</div>
<section id="writing-and-running-your-own-scripts" class="level3">
<h3 class="anchored" data-anchor-id="writing-and-running-your-own-scripts">Writing and running your own scripts</h3>
<p>In its most basic form, a shell script is simply a text file containing a list of commands that is run in sequence from top to bottom. This kind of script can be run by providing it as an argument to the <code>bash</code> command, as in the following example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> date</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Thu</span> 12 Nov 14:21:16 GMT 2020</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> echo hello</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">hello</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> ls /homes/genomes/mouse/</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">GRCm38</span>  mm10  mm9  UCSC</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> echo date <span class="op">&gt;</span> my_script.sh</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> echo <span class="st">'echo hello'</span> <span class="op">&gt;&gt;</span> my_script.sh</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> echo <span class="st">'ls /homes/genomes/mouse'</span> <span class="op">&gt;&gt;</span> my_script.sh</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> cat my_script.sh</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> hello</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> /homes/genomes/mouse/</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> bash my_script.sh</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="ex">Thu</span> 12 Nov 14:26:29 GMT 2020</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="ex">hello</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="ex">GRCm38</span>  mm10  mm9  UCSC</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is also possible to create a script that can be run as a standalone program rather than as an argument to <code>bash</code>. This involves two steps:</p>
<ol type="1">
<li>Add the line <code>#!/bin/bash</code> (known as a <a href="https://en.wikipedia.org/wiki/Shebang_(Unix)">shebang</a> line) as the first line of the file using a text editor</li>
<li>Use the <code>chmod</code> command to make the file executable
<ul>
<li><code>chmod a+x my_script.sh</code> will change the permissions on the script so that any user can run it</li>
</ul></li>
</ol>
<p>Once these steps have been followed, it will be possible to execute the script using its path:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> cat my_script.sh</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> hello</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> /homes/genomes/mouse/</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> ls <span class="at">-l</span> ./my_script.sh</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="ex">-rwxrwxr-x</span> 1 [USERNAME] [USERNAME] 0 Nov 13 21:40 my_script.sh</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> ./my_script.sh</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="ex">Thu</span> 12 Nov 14:26:29 GMT 2020</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="ex">hello</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="ex">GRCm38</span>  mm10  mm9  UCSC</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~$</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Note:</em> If you want to be able to run your script just by typing its name, you need to move it to a directory that is included in the <em>$PATH</em> environment variable.</p>
<ul>
<li>To check which folders are listed in <em>$PATH</em>, you can type <code>echo $PATH</code></li>
<li>To add a directory to <em>$PATH</em> permanently, add the line <code>export PATH=[YOUR DIRECTORY]:$PATH</code> to the end of the <code>~/.bashrc</code> file, then run the command <code>source ~/.bashrc</code></li>
<li>If you run a program from a directory in <em>$PATH</em>, it can be useful to check the full path to that program to make sure you’re not inadvertently running another program with the same name. You can do this by using the <code>which</code> command.</li>
</ul>
</section>
<section id="editing-scripts" class="level3">
<h3 class="anchored" data-anchor-id="editing-scripts">Editing scripts</h3>
<p>You can edit your scripts using a text editor such as <em>emacs</em> or <em>vim</em>.</p>
<ul>
<li><code>emacs</code> opens emacs
<ul>
<li>you can then take the tutorial by typing Ctrl+h t</li>
</ul></li>
<li><code>vim</code> opens vim
<ul>
<li>you can take a tutorial by running the <code>vimtutor</code> command at the bash prompt</li>
</ul></li>
</ul>
</section>
</section>
<section id="reproducible-data-analysis-with-scripts" class="level2">
<h2 class="anchored" data-anchor-id="reproducible-data-analysis-with-scripts">Reproducible data analysis with scripts</h2>
<div class="key-points">
<h2 class="anchored">
<i class="fas fa-thumbtack"></i> Key points
</h2>
<ul>
<li>Data analysis should be reproducible
<ul>
<li>You should be able to recreate all of the steps in your analysis</li>
<li>Other researchers should also be able to recreate your analysis to verify your results</li>
</ul></li>
<li>The bash shell makes it easy to analyse data interactively. However, if data is analysed this way it can be difficult to keep track of exactly which steps were taken to produce a given result</li>
<li>This problem can be solved by creating a script that contains all of the commands needed to produce the results
<ul>
<li>You and other researchers can then recreate your analysis by running the script</li>
</ul></li>
<li>This section presents a bash script that replicates the case study performed in the previous section</li>
</ul>
</div>
<p>In the previous section we showed how to perform a simple analysis of some RNA-Seq data from scratch. In order turn this into a reproducible analysis workflow, it is necessary to put all of the commands that make up the analysis into a script.</p>
<section id="creating-a-simple-data-analysis-script" class="level3">
<h3 class="anchored" data-anchor-id="creating-a-simple-data-analysis-script">Creating a simple data analysis script</h3>
<p>The simplest way to create a script is to put all of the commands that made up the analysis into a file in the order in which they were run, creating a file that looks like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> analysis</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> analysis</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 00_source_data</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 00_source_data</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 01_star_index</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 01_star_index</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">nice</span> STAR <span class="at">--runThreadN</span> 5 <span class="at">--runMode</span> genomeGenerate <span class="at">--genomeDir</span> . <span class="at">--genomeFastaFiles</span> ./yeast_genome.fasta <span class="at">--genomeSAindexNbases</span> 10</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 02_aligned_reads</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 02_aligned_reads</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">nice</span> STAR <span class="at">--genomeDir</span> ../01_star_index/ <span class="at">--readFilesIn</span> ../00_source_data/raw_yeast_rnaseq_data.fastq <span class="at">--runThreadN</span> 5 <span class="at">--outFileNamePrefix</span> raw_yeast_rnaseq_data. <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 03_coverage</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 03_coverage</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-ibam</span> ../02_aligned_reads/raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam <span class="op">&gt;</span> raw_yeast_rnaseq_data.genomecov.bg</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 04_gene_overlap_counts</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 04_gene_overlap_counts</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-c</span> <span class="at">-a</span> yeast_genes.fixed.bed <span class="at">-b</span> ../02_aligned_reads/raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam <span class="kw">|</span> <span class="fu">awk</span> <span class="at">-F</span><span class="st">'\t'</span> <span class="st">'$7&gt;0'</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-k7,7nr</span> <span class="op">&gt;</span> raw_yeast_overlap_data.gene_overlap_counts.bed</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-10</span> 04_gene_overlap_counts/raw_yeast_overlap_data.gene_overlap_counts.bed</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that this works by running the above file (which is saved on the server as ‘/library/training/bioinformatics_on_the_command_line/analysis_raw.sh’):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> ls</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">analysis</span>  bioinformatics_on_the_command_line_files  bioinformatics_on_the_command_line_files.tar.gz</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> mv analysis interactive_analysis</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> ls</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bioinformatics_on_the_command_line_files</span>  bioinformatics_on_the_command_line_files.tar.gz  interactive_analysis</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> bash /library/training/bioinformatics_on_the_command_line/analysis_raw.sh</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="bu">.</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> 00_source_data</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> raw_yeast_rnaseq_data.fastq <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="ex">1</span> directory, 1 file</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:45 ..... started STAR run</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:45 ... starting to generate Genome files</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:46 ... starting to sort Suffix Array. This may take a long time...</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:46 ... sorting Suffix Array chunks and saving them to disk...</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:54 ... loading chunks from disk, packing SA...</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:55 ... finished generating suffix array</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:55 ... generating Suffix Array index</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:56 ... completed Suffix Array index</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:56 ... writing Genome to disk ...</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:56 ... writing Suffix Array to disk ...</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:57 ... writing SAindex to disk</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:57 ..... finished successfully</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="bu">.</span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 00_source_data</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── raw_yeast_rnaseq_data.fastq <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> 01_star_index</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> chrLength.txt</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> chrNameLength.txt</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> chrName.txt</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> chrStart.txt</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> Genome</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> genomeParameters.txt</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> Log.out</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> SA</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> SAindex</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> yeast_genome.fasta <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a><span class="ex">2</span> directories, 11 files</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:57 ..... started STAR run</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:57 ..... loading genome</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:23:57 ..... started mapping</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:24:00 ..... finished mapping</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:24:00 ..... started sorting BAM</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:24:00 ..... finished successfully</span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a><span class="bu">.</span></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 00_source_data</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── raw_yeast_rnaseq_data.fastq <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 01_star_index</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrLength.txt</span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrNameLength.txt</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrName.txt</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrStart.txt</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── Genome</span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── genomeParameters.txt</span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── Log.out</span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SA</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SAindex</span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── yeast_genome.fasta <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta</span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> 02_aligned_reads</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam</span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> raw_yeast_rnaseq_data.Log.final.out</span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> raw_yeast_rnaseq_data.Log.out</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> raw_yeast_rnaseq_data.Log.progress.out</span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> raw_yeast_rnaseq_data.SJ.out.tab</span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> raw_yeast_rnaseq_data._STARtmp</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>        <span class="ex">└──</span> BAMsort</span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> 0</span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> 1</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> 2</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> 3</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>            <span class="ex">└──</span> 4</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a><span class="ex">10</span> directories, 16 files</span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a><span class="bu">.</span></span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 00_source_data</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── raw_yeast_rnaseq_data.fastq <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq</span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 01_star_index</span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrLength.txt</span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrNameLength.txt</span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrName.txt</span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrStart.txt</span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── Genome</span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── genomeParameters.txt</span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── Log.out</span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SA</span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SAindex</span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── yeast_genome.fasta <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta</span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 02_aligned_reads</span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam</span>
<span id="cb4-91"><a href="#cb4-91" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.Log.final.out</span>
<span id="cb4-92"><a href="#cb4-92" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.Log.out</span>
<span id="cb4-93"><a href="#cb4-93" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.Log.progress.out</span>
<span id="cb4-94"><a href="#cb4-94" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.SJ.out.tab</span>
<span id="cb4-95"><a href="#cb4-95" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── raw_yeast_rnaseq_data._STARtmp</span>
<span id="cb4-96"><a href="#cb4-96" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     └── BAMsort</span>
<span id="cb4-97"><a href="#cb4-97" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         ├── 0</span>
<span id="cb4-98"><a href="#cb4-98" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         ├── 1</span>
<span id="cb4-99"><a href="#cb4-99" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         ├── 2</span>
<span id="cb4-100"><a href="#cb4-100" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         ├── 3</span>
<span id="cb4-101"><a href="#cb4-101" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         └── 4</span>
<span id="cb4-102"><a href="#cb4-102" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> 03_coverage</span>
<span id="cb4-103"><a href="#cb4-103" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> raw_yeast_rnaseq_data.genomecov.bg</span>
<span id="cb4-104"><a href="#cb4-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-105"><a href="#cb4-105" aria-hidden="true" tabindex="-1"></a><span class="ex">11</span> directories, 17 files</span>
<span id="cb4-106"><a href="#cb4-106" aria-hidden="true" tabindex="-1"></a><span class="bu">.</span></span>
<span id="cb4-107"><a href="#cb4-107" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 00_source_data</span>
<span id="cb4-108"><a href="#cb4-108" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── raw_yeast_rnaseq_data.fastq <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq</span>
<span id="cb4-109"><a href="#cb4-109" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 01_star_index</span>
<span id="cb4-110"><a href="#cb4-110" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrLength.txt</span>
<span id="cb4-111"><a href="#cb4-111" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrNameLength.txt</span>
<span id="cb4-112"><a href="#cb4-112" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrName.txt</span>
<span id="cb4-113"><a href="#cb4-113" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── chrStart.txt</span>
<span id="cb4-114"><a href="#cb4-114" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── Genome</span>
<span id="cb4-115"><a href="#cb4-115" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── genomeParameters.txt</span>
<span id="cb4-116"><a href="#cb4-116" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── Log.out</span>
<span id="cb4-117"><a href="#cb4-117" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SA</span>
<span id="cb4-118"><a href="#cb4-118" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── SAindex</span>
<span id="cb4-119"><a href="#cb4-119" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── yeast_genome.fasta <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta</span>
<span id="cb4-120"><a href="#cb4-120" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 02_aligned_reads</span>
<span id="cb4-121"><a href="#cb4-121" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam</span>
<span id="cb4-122"><a href="#cb4-122" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.Log.final.out</span>
<span id="cb4-123"><a href="#cb4-123" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.Log.out</span>
<span id="cb4-124"><a href="#cb4-124" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.Log.progress.out</span>
<span id="cb4-125"><a href="#cb4-125" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_yeast_rnaseq_data.SJ.out.tab</span>
<span id="cb4-126"><a href="#cb4-126" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── raw_yeast_rnaseq_data._STARtmp</span>
<span id="cb4-127"><a href="#cb4-127" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     └── BAMsort</span>
<span id="cb4-128"><a href="#cb4-128" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         ├── 0</span>
<span id="cb4-129"><a href="#cb4-129" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         ├── 1</span>
<span id="cb4-130"><a href="#cb4-130" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         ├── 2</span>
<span id="cb4-131"><a href="#cb4-131" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         ├── 3</span>
<span id="cb4-132"><a href="#cb4-132" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         └── 4</span>
<span id="cb4-133"><a href="#cb4-133" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> 03_coverage</span>
<span id="cb4-134"><a href="#cb4-134" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── raw_yeast_rnaseq_data.genomecov.bg</span>
<span id="cb4-135"><a href="#cb4-135" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> 04_gene_overlap_counts</span>
<span id="cb4-136"><a href="#cb4-136" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> raw_yeast_overlap_data.gene_overlap_counts.bed</span>
<span id="cb4-137"><a href="#cb4-137" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> yeast_genes.fixed.bed <span class="at">-</span><span class="op">&gt;</span> ../../bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed</span>
<span id="cb4-138"><a href="#cb4-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-139"><a href="#cb4-139" aria-hidden="true" tabindex="-1"></a><span class="ex">12</span> directories, 19 files</span>
<span id="cb4-140"><a href="#cb4-140" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 460922  466869  RDN37-2 .   <span class="at">-</span>   3730</span>
<span id="cb4-141"><a href="#cb4-141" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 451785  457732  RDN37-1 .   <span class="at">-</span>   3582</span>
<span id="cb4-142"><a href="#cb4-142" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 468812  468931  RDN5-2  .   +   3063</span>
<span id="cb4-143"><a href="#cb4-143" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 468826  468958  YLR154C-H   .   <span class="at">-</span>   3063</span>
<span id="cb4-144"><a href="#cb4-144" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 472464  472583  RDN5-3  .   +   3060</span>
<span id="cb4-145"><a href="#cb4-145" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 472478  472610  YLR156C-A   .   <span class="at">-</span>   3060</span>
<span id="cb4-146"><a href="#cb4-146" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 482044  482163  RDN5-4  .   +   3055</span>
<span id="cb4-147"><a href="#cb4-147" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 482058  482190  YLR157C-C   .   <span class="at">-</span>   3055</span>
<span id="cb4-148"><a href="#cb4-148" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 485696  485815  RDN5-5  .   +   3052</span>
<span id="cb4-149"><a href="#cb4-149" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 485710  485842  YLR159C-A   .   <span class="at">-</span>   3052</span>
<span id="cb4-150"><a href="#cb4-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-151"><a href="#cb4-151" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="creating-an-improved-data-analysis-script" class="level3">
<h3 class="anchored">Creating an improved data analysis script</h3>
<p>In this section we improve our script by doing the following:</p>
<ul>
<li>Adding a shebang line to the script so that it can be run as a standalone program</li>
<li>Ensuring that the script does not keep running if one of the commands within it fails
<ul>
<li>This is what the <code>set -euo pipefail</code> command at the top of the script does</li>
</ul></li>
<li>Deleting unnecessary commands</li>
<li>Adding spacing and comments to make the script easier to read
<ul>
<li>In bash, lines starting with <code>#</code> are taken to be comments</li>
</ul></li>
<li>Specifying inputs as variables in order to make the script more general and easier to maintain, and using paths relative to ‘~’ rather than the current working directory so that the script can be run in any directory and will still find the files</li>
<li>Adding a few lines to the script to create a run report, which specifies the software versions used for the analysis</li>
</ul>
<p>The updated script is shown here:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This is simple RNA-Seq analysis workflow that does the following:</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># - Creates a STAR index for the genome fasta file specified in the variable GENOME_FASTA</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># - Aligns the raw reads specified in the fastq file referenced by the variable RAW_FASTQ</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># - Computes the genome coverage of the aligned reads, producing a bedGraph file</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co"># - Counts the overlaps over the genes specified in the file referenced by the variable GENES_BED</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Global variables ----</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="va">RAW_FASTQ</span><span class="op">=</span><span class="st">'~/course/bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq'</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="va">GENOME_FASTA</span><span class="op">=</span><span class="st">'~/course/bioinformatics_on_the_command_line_files/yeast_genome.fasta'</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="va">GENES_BED</span><span class="op">=</span><span class="st">'~/course/bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed'</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="st">"</span><span class="va">$RAW_FASTQ</span><span class="st">"</span> <span class="st">"</span><span class="va">$GENOME_FASTA</span><span class="st">"</span> <span class="st">"</span><span class="va">$GENES_BED</span><span class="st">"</span> <span class="op">&gt;</span> /dev/null</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="va">RAW_FASTQ_BASENAME_PREFIX</span><span class="op">=</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$RAW_FASTQ</span> .fastq<span class="kw">`</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="va">GENOME_FASTA_BASENAME</span><span class="op">=</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$GENOME_FASTA</span><span class="kw">`</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="va">THREADS</span><span class="op">=</span>5</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Pipeline commands ----</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a directory for the results of the analysis</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> analysis</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> analysis</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Link to the fastq containing the raw sequences in '00_source_data'</span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 00_source_data</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 00_source_data</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> <span class="st">"</span><span class="va">$RAW_FASTQ</span><span class="st">"</span></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a STAR index for the genome fasta file, storing all of the results in '01_star_index'</span></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'Creating STAR index...'</span></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 01_star_index</span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 01_star_index</span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> <span class="st">"</span><span class="va">$GENOME_FASTA</span><span class="st">"</span></span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a><span class="fu">nice</span> STAR <span class="at">--runThreadN</span> <span class="st">"</span><span class="va">$THREADS</span><span class="st">"</span> <span class="at">--runMode</span> genomeGenerate <span class="at">--genomeDir</span> . <span class="at">--genomeFastaFiles</span> <span class="st">"</span><span class="va">$GENOME_FASTA_BASENAME</span><span class="st">"</span> <span class="at">--genomeSAindexNbases</span> 10</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Align the raw sequences using STAR, storing all of the results in '02_aligned_reads'</span></span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'Aligning raw reads...'</span></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 02_aligned_reads</span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 02_aligned_reads</span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a><span class="fu">nice</span> STAR <span class="at">--genomeDir</span> ../01_star_index/ <span class="at">--readFilesIn</span> ../00_source_data/<span class="st">"</span><span class="va">$RAW_FASTQ_BASENAME_PREFIX</span><span class="st">"</span>.fastq <span class="at">--runThreadN</span> <span class="st">"</span><span class="va">$THREADS</span><span class="st">"</span> <span class="at">--outFileNamePrefix</span> <span class="st">"</span><span class="va">$RAW_FASTQ_BASENAME_PREFIX</span><span class="st">"</span>. <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a genome coverage file in bedGraph format, and store it in '03_coverage'</span></span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'Creating genome coverage file...'</span></span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 03_coverage</span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 03_coverage</span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-ibam</span> ../02_aligned_reads/<span class="st">"</span><span class="va">$RAW_FASTQ_BASENAME_PREFIX</span><span class="st">"</span>.Aligned.sortedByCoord.out.bam <span class="op">&gt;</span> <span class="st">"</span><span class="va">$RAW_FASTQ_BASENAME_PREFIX</span><span class="st">"</span>.genomecov.bg</span>
<span id="cb5-65"><a href="#cb5-65" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-66"><a href="#cb5-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-67"><a href="#cb5-67" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the gene overlap counts, and store them as a bed file in '04_gene_overlap_counts'</span></span>
<span id="cb5-68"><a href="#cb5-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-69"><a href="#cb5-69" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'Computing gene overlap counts...'</span></span>
<span id="cb5-70"><a href="#cb5-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-71"><a href="#cb5-71" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 04_gene_overlap_counts</span>
<span id="cb5-72"><a href="#cb5-72" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 04_gene_overlap_counts</span>
<span id="cb5-73"><a href="#cb5-73" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed</span>
<span id="cb5-74"><a href="#cb5-74" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-c</span> <span class="at">-a</span> <span class="st">"</span><span class="va">$GENES_BED</span><span class="st">"</span> <span class="at">-b</span> ../02_aligned_reads/<span class="st">"</span><span class="va">$RAW_FASTQ_BASENAME_PREFIX</span><span class="st">"</span>.Aligned.sortedByCoord.out.bam <span class="kw">|</span> <span class="fu">awk</span> <span class="at">-F</span><span class="st">'\t'</span> <span class="st">'$7&gt;0'</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-k7,7nr</span> <span class="op">&gt;</span> <span class="st">"</span><span class="va">$RAW_FASTQ_BASENAME_PREFIX</span><span class="st">"</span>.gene_overlap_counts.bed</span>
<span id="cb5-75"><a href="#cb5-75" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-76"><a href="#cb5-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-77"><a href="#cb5-77" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a run report</span></span>
<span id="cb5-78"><a href="#cb5-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-79"><a href="#cb5-79" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="kw">`</span><span class="fu">realpath</span> <span class="va">$0</span><span class="kw">`</span><span class="st">" run completed successfully."</span> <span class="op">&gt;</span> run_report.txt</span>
<span id="cb5-80"><a href="#cb5-80" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-81"><a href="#cb5-81" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'Software versions:'</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-82"><a href="#cb5-82" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'STAR'</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-83"><a href="#cb5-83" aria-hidden="true" tabindex="-1"></a><span class="ex">STAR</span> <span class="at">--version</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-84"><a href="#cb5-84" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'bedtools'</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-85"><a href="#cb5-85" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> <span class="at">--version</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/^bedtools //'</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-86"><a href="#cb5-86" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'sort'</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-87"><a href="#cb5-87" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span> <span class="at">--version</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/^sort //'</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-1</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-88"><a href="#cb5-88" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'awk'</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-89"><a href="#cb5-89" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="at">--version</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-1</span> <span class="op">&gt;&gt;</span> run_report.txt</span>
<span id="cb5-90"><a href="#cb5-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-91"><a href="#cb5-91" aria-hidden="true" tabindex="-1"></a><span class="co"># If we've got here the pipeline has completed successfully</span></span>
<span id="cb5-92"><a href="#cb5-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-93"><a href="#cb5-93" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'Pipeline completed successfully.'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Again, we can see that this works by running the above file (which is saved on the server as ‘/library/training/bioinformatics_on_the_command_line/analysis_improved.sh’):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> ls</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">analysis</span>  bioinformatics_on_the_command_line_files  bioinformatics_on_the_command_line_files.tar.gz interactive_analysis</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> mv analysis analysis_raw</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> ls</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="ex">analysis_raw</span>                              bioinformatics_on_the_command_line_files.tar.gz</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="ex">bioinformatics_on_the_command_line_files</span>  interactive_analysis</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> bash /library/training/bioinformatics_on_the_command_line/analysis_improved.sh</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="ex">Creating</span> STAR index...</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:09 ..... started STAR run</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:09 ... starting to generate Genome files</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:10 ... starting to sort Suffix Array. This may take a long time...</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:10 ... sorting Suffix Array chunks and saving them to disk...</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:18 ... loading chunks from disk, packing SA...</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:19 ... finished generating suffix array</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:19 ... generating Suffix Array index</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:20 ... completed Suffix Array index</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:20 ... writing Genome to disk ...</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:20 ... writing Suffix Array to disk ...</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:21 ... writing SAindex to disk</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:21 ..... finished successfully</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="ex">Aligning</span> raw reads...</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:21 ..... started STAR run</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:21 ..... loading genome</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:21 ..... started mapping</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:24 ..... finished mapping</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:24 ..... started sorting BAM</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a><span class="ex">Nov</span> 16 16:38:24 ..... finished successfully</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a><span class="ex">Creating</span> genome coverage file...</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a><span class="ex">Computing</span> gene overlap counts...</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a><span class="ex">Pipeline</span> completed successfully.</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> cat analysis/run_report.txt</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a><span class="ex">/library/training/bioinformatics_on_the_command_line/analysis_improved.sh</span> run completed successfully.</span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a><span class="ex">Mon</span> 16 Nov 16:38:36 GMT 2020</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a><span class="ex">Software</span> versions:</span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a><span class="ex">STAR</span></span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a><span class="ex">2.7.3a</span></span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span></span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a><span class="ex">v2.27.0</span></span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span></span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a><span class="kw">(</span><span class="ex">GNU</span> coreutils<span class="kw">)</span> <span class="ex">8.28</span></span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span></span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true" tabindex="-1"></a><span class="ex">GNU</span> Awk 4.1.4, API: 1.1 <span class="er">(</span><span class="ex">GNU</span> MPFR 4.0.1, GNU MP 6.1.2<span class="kw">)</span></span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core2:~/course$</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="challenge">
<h2 class="anchored" data-anchor-id="creating-an-improved-data-analysis-script">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>What is the line <code>ls "$RAW_FASTQ" "$GENOME_FASTA" "$GENES_BED" &gt; /dev/null</code> in the script for?</p>
<details>
<summary>
Solution
</summary>
<div class="solution proof">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<p>If the files referenced by <em>RAW_FASTQ</em>, <em>GENOME_FASTA</em>, and <em>GENES_BED</em> exist, the line does nothing, as the output of the command is redirected to <code>/dev/null</code> (which basically means that it is ignored). However, if any of the files don’t exist this command will fail with an error message. Because we have included the line <code>set -eou pipefail</code> this will cause the script to stop running, which is what we want if one or more of the output files is missing.</p>
</div>
</details>
</div>
</section>
<section id="managing-your-scripts" class="level3">
<h3 class="anchored" data-anchor-id="managing-your-scripts">Managing your scripts</h3>
<p>Once you start writing scripts, it is a good idea to use a <em>version control system</em> to keep track of the changes you make to your scripts. A good choice for this would be <code>git</code>. The following five commands will allow you to use <code>git</code> for version control:</p>
<ul>
<li><code>git init</code> to initialise a git repository in the current working directory</li>
<li><code>git add -A; git commit -m "Latest updates"</code> to commit any changes to files in the current working directory to the repository</li>
<li><code>git status</code> to see if any changes have been made since the last commit</li>
<li><code>git diff</code> to see the differences between the files in the current working directory and the last commit</li>
<li><code>git show HEAD:[NAME OF FILE IN GIT]</code> to view the most recent version of a file in the git repository
<ul>
<li>You can save the output of <code>git show</code> to a file by redirecting its output using <code>&gt;</code></li>
</ul></li>
</ul>
<p>To learn more about <code>git</code>, you can look at this <a href="https://swcarpentry.github.io/git-novice/">software carpentry course</a>, which covers it in a lot more detail.</p>
<p><em>Note:</em> <code>git</code> should only be used for small text files such as scripts. It is not designed to work with large data files.</p>
</section>
</section>
<section id="closing-thoughts-practical-workflows-on-the-command-line" class="level2">
<h2 class="anchored" data-anchor-id="closing-thoughts-practical-workflows-on-the-command-line">Closing thoughts: practical workflows on the command line</h2>
<p>In this section we created a data analysis script in bash that satisfies two major requirements for reproducible data analysis:</p>
<ul>
<li>The script provides a complete and accurate record of the steps that were taken to produce the output files, along with a record of the versions of the software tools used to manipulate the data</li>
<li>The script is also in a form that would be easy to share with other researchers, allowing them to replicate your analysis easily</li>
</ul>
<p>The main drawback of using bash scripts to represent workflows is that they are often impractical. The example presented here uses an extremely small input FASTQ file, and an organism with a relatively small genome. As a result it can be run from scratch in under a minute. Running the same pipeline with a realistically sized input FASTQ file and a larger genome could take hours to run, so it is no longer practical to re-run the pipeline every time you make a change. A workaround for this would be to comment out parts of the script that you do not want to re-run, however this is considered bad practice as it introduces the possibility of human error in selecting the parts of the pipeline that need to be re-run when the pipeline is updated.</p>
<p>If you are writing analysis pipelines, you should look into using a modern workflow system such as <a href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a> or <a href="https://www.nextflow.io/">Nextflow</a>. Workflows written for these workflow systems can be run repeatedly, and the workflow system will automatically work out which steps of the analysis workflow need to be re-run based on the timestamps of the files. They also provide a number of other useful features:</p>
<ul>
<li>They work out which steps of the analysis workflow can be run in parallel, and assign steps to different cores when possible, which can speed up the analysis considerably</li>
<li>They simplify the process of running workflows on a computing cluster</li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>